### ISSUE TITLE
Out of memory

### ISSUE THREAD

**Issue by OlafvdSpek** [phpmyadmin/phpmyadmin#16734]
pMA 5.1.1-dev / PHP 8

When trying to browse a large table.. 

```
PHP Fatal error:  Allowed memory size of 134217728 bytes exhausted (tried to allocate 68502032 bytes) in libraries/classes/Dbal/DbiMysqli.php on line 244

```

```php
    public function realQuery($query, $mysqli, $options)
    {
        if ($options == ($options | DatabaseInterface::QUERY_STORE)) {
            $method = MYSQLI_STORE_RESULT;
        } elseif ($options == ($options | DatabaseInterface::QUERY_UNBUFFERED)) {
            $method = MYSQLI_USE_RESULT;
        } else {
            $method = 0;
        }

        return $mysqli->query($query, $method); // 244
    }
```

`($options == ($options | DatabaseInterface::QUERY_STORE))` also looks very curious.

```
2021-03-14 09:37:51: SET collation_connection = 'utf8mb4_unicode_ci';
2021-03-14 09:37:51: SHOW SESSION VARIABLES LIKE 'FOREIGN_KEY_CHECKS';
2021-03-14 09:37:51: SHOW SESSION VARIABLES LIKE 'FOREIGN_KEY_CHECKS';
2021-03-14 09:37:51: SHOW SESSION VARIABLES LIKE 'FOREIGN_KEY_CHECKS';
2021-03-14 09:37:51: SHOW SESSION VARIABLES LIKE 'FOREIGN_KEY_CHECKS';
2021-03-14 09:37:51: SHOW SESSION VARIABLES LIKE 'FOREIGN_KEY_CHECKS';
2021-03-14 09:37:51: SELECT DATABASE()
2021-03-14 09:37:51: select * from users where uid != 0 LIMIT 0, 100 
2021-03-14 09:37:51: select * from users where uid != 0    
2021-03-14 09:37:51: SHOW WARNINGS
```

The second query without limit is the culprit..
The repeated FOREIGN_KEY_CHECKS query looks sub-optimal too.

**Comment by williamdes** [comment]
I could not confirm this, even using a Docker container with very little RAM
Could you help reproducing this issue ?

**Comment by OlafvdSpek** [comment]
Are you using PHP 8.0 with memory_limit = 128M (the default)? The amount of RAM shouldn't matter.
How did you enable the query log? Do you see the second query without limit?

**Comment by williamdes** [comment]
I am trying to debug this on QA but for now did not find any evidence.
Could you apply manually this patch ? (that will help understand the process)
And maybe post the logs here afterwards please :)
```diff
diff --git a/libraries/classes/DatabaseInterface.php b/libraries/classes/DatabaseInterface.php
index d05d86aee0..eafff3db6f 100644
--- a/libraries/classes/DatabaseInterface.php
+++ b/libraries/classes/DatabaseInterface.php
@@ -237,9 +237,15 @@ class DatabaseInterface implements DbalInterface
 
                 syslog(
                     LOG_INFO,
-                    'SQL[' . basename($_SERVER['SCRIPT_NAME']) . ']: '
+                    'SQL[' . basename($_SERVER['SCRIPT_NAME']) . '?route=' . Routing::getCurrentRoute() . ']: '
                     . sprintf('%0.3f', $time) . '(W:' . $warningsCount . ') > ' . $query
                 );
                 closelog();
             }
         }
```

**Comment by OlafvdSpek** [comment]
`route=/table/search`

**Comment by williamdes** [comment]
Here is my current hack, they are hard to find but should be in a `/var/log`.
If you apply it, that should help debug even more. I added `C` for cached or not
```php

                syslog(
                    LOG_INFO,
                    sprintf(
                        'SQL[%s?route=%s]: %0.3f(W:%s,C:%s,L:0x%02X) > %s',
                        basename($_SERVER['SCRIPT_NAME']),
                        Routing::getCurrentRoute(),
                        $time,
                        $warningsCount,
                        $cache_affected_rows ? 'y' : 'n',
                        $link,
                        $query
                    )
                );
                error_log(
                    sprintf(
                        'SQL[%s?route=%s]: %0.3f(W:%s,C:%s,L:0x%02X) > %s',
                        basename($_SERVER['SCRIPT_NAME']),
                        Routing::getCurrentRoute(),
                        $time,
                        $warningsCount,
                        $cache_affected_rows ? 'y' : 'n',
                        $link,
                        $query
                    ) . PHP_EOL,
                    3,
                    ROOT_PATH . '/build/sql-log.txt'
                );
```

**Comment by OlafvdSpek** [comment]
`if ($GLOBALS['cfg']['DBG']['sqllog']) {` ;)

**Comment by williamdes** [comment]
I have only one fix in mind that could cause such an issue, and it is the issue #15652
If you revert #16470 does it work ?

**Comment by OlafvdSpek** [comment]
```
                $statement = $analyzed_sql_results['statement'];
                $token_list = $analyzed_sql_results['parser']->list;
                $replaces = [
                    // Remove ORDER BY to decrease unnecessary sorting time
                    [   
                        'ORDER BY',
                        '',
                    ],
                    // Removes LIMIT clause that might have been added
                    [   
                        'LIMIT',
                        '',
                    ],
                ];
                $count_query = Query::replaceClauses(
                    $statement,
                    $token_list,
                    $replaces
                );
                $unlim_num_rows = $this->dbi->numRows($this->dbi->tryQuery($count_query));
```

**Comment by williamdes** [comment]
I did some updates to my deploy container to be able to use it locally
https://github.com/sudo-bot/gh-deployer-container#skip-deploy-and-use-your-workspace-folder

I think I have reproduced it, depending on my research criteria it crashes or not the fpm thread

**Comment by williamdes** [comment]
Reverting the #16470 fixes the issue

cc @ibennetch this issue is release blocking until fixed, working on it

**Comment by williamdes** [comment]
Hi @OlafvdSpek 
I think I have a fix, could you try it please ?
```diff
diff --git a/libraries/classes/Sql.php b/libraries/classes/Sql.php
index ed83e20174..fecd059af5 100644
--- a/libraries/classes/Sql.php
+++ b/libraries/classes/Sql.php
@@ -803,6 +803,11 @@ class Sql
                 $statement = $analyzed_sql_results['statement'];
                 $token_list = $analyzed_sql_results['parser']->list;
                 $replaces = [
+                    // Replace select by COUNT
+                    [
+                        'SELECT',
+                        'SELECT COUNT(*)',
+                    ],
                     // Remove ORDER BY to decrease unnecessary sorting time
                     [
                         'ORDER BY',
@@ -819,7 +824,7 @@ class Sql
                     $token_list,
                     $replaces
                 );
-                $unlim_num_rows = $this->dbi->numRows($this->dbi->tryQuery($count_query));
+                $unlim_num_rows = $this->dbi->fetchValue($count_query);
             }
         } else {// not $is_select
             $unlim_num_rows = 0;
```

**Comment by OlafvdSpek** [comment]
If the original query is like `select count(*) from users`, the result set size is 1, but wouldn't your query return something different?

**Comment by williamdes** [comment]
> If the original query is like `select count(*) from users`, the result set size is 1, but wouldn't your query return something different?

It should not, the replacing will give the same query as the input and count

The snapshots are up to date, let me know if you find something I missed

**Comment by OlafvdSpek** [comment]
https://demo.phpmyadmin.net/master-config/index.php?route=/sql&db=mysql&table=help_relation&pos=0&server=1

![image](https://user-images.githubusercontent.com/2218936/115985255-488b2000-a5ab-11eb-98be-48870c4583c8.png)

**Comment by williamdes** [comment]
Should be: (4.9 output)
![Screenshot_20210425_101731_com brave browser](https://user-images.githubusercontent.com/7784660/115986250-9efa5d80-a5af-11eb-8709-4ec479dbc12a.jpg)
![Screenshot_20210425_101209_com brave browser](https://user-images.githubusercontent.com/7784660/115986255-a3267b00-a5af-11eb-94c3-2b2a4089f630.jpg)

**Comment by williamdes** [comment]
I think I have an idea on how to patch this
```sql
SELECT COUNT(*) FROM (SELECT * from help_category group by help_category_id) as cnt
```
The would give the maximum rows in an easy way
As far as I can think that would be a solution for all queries

**Comment by williamdes** [comment]
@ibennetch could you give some advice on this one please ?

**Comment by williamdes** [comment]
```diff
diff --git a/libraries/classes/Sql.php b/libraries/classes/Sql.php
index 8cde36c757..6c94495f18 100644
--- a/libraries/classes/Sql.php
+++ b/libraries/classes/Sql.php
@@ -803,11 +803,6 @@ class Sql
                 $statement = $analyzed_sql_results['statement'];
                 $token_list = $analyzed_sql_results['parser']->list;
                 $replaces = [
-                    // Replace select by COUNT
-                    [
-                        'SELECT',
-                        'SELECT COUNT(*)',
-                    ],
                     // Remove ORDER BY to decrease unnecessary sorting time
                     [
                         'ORDER BY',
@@ -819,11 +814,11 @@ class Sql
                         '',
                     ],
                 ];
-                $count_query = Query::replaceClauses(
+                $count_query = 'SELECT COUNT(*) FROM (' . Query::replaceClauses(
                     $statement,
                     $token_list,
                     $replaces
-                );
+                ) . ') as count';
                 $unlim_num_rows = $this->dbi->fetchValue($count_query);
                 if ($unlim_num_rows === false) {
                     $unlim_num_rows = 0;
```

I applied this and it works fine

**Comment by williamdes** [comment]
@ibennetch this is the last issue blocking the release, what do you think about my patch ?

**Comment by williamdes** [comment]
Should be fixed by #16902
